<a href="https://colab.research.google.com/github/ashleyak7/MSCI-151-week-1-workshops/blob/main/CW4_AAK.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

excel_file_path = '/content/sample_data/MSCI151 Data.xlsx'
df = pd.read_excel(excel_file_path)
print(df.columns)

Index(['id', 'Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size',
       'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender'],
      dtype='object')


SyntaxError: invalid syntax (ipython-input-558758558.py, line 1)

# Task
Perform a predictive analysis using a Random Forest model on the data in "data/content/sample_data/MSCI151 Data.xlsx". The target variable is 'Profitability', and the features are 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO Education', and 'CEO Gender'. Split the data into training and testing sets using an 80/20 ratio.

## Prepare data

### Subtask:
Define the features (X) and the target variable (y) from the DataFrame `df`.


**Reasoning**:
Define the feature columns and the target variable from the dataframe.



In [5]:
feature_cols = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender']
X = df[feature_cols]
y = df['Profitability']

## Split data

### Subtask:
Split the data into training and testing sets using an 80/20 rule.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train model

### Subtask:
Train a Random Forest Regressor model on the training data.


**Reasoning**:
Train a Random Forest Regressor model on the training data.



In [7]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [9]:
import pandas as pd

excel_file_path = '/content/sample_data/MSCI151 coursework 4 Data.new.xlsx'
df_new = pd.read_excel(excel_file_path)
print(df_new.columns)

Index(['id', 'Profitability', 'debt', 'RnD', 'Firm Age', 'Firm Size',
       'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender',
       'Marketing', 'Product_sentiment', 'Politicans', 'PoB'],
      dtype='object')


## Evaluate model

### Subtask:
Evaluate the trained model on the testing data using appropriate metrics.


**Reasoning**:
Import necessary evaluation metrics, make predictions on the test set, calculate the metrics, and print them.



In [8]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

y_pred = model.predict(X_test)

mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"R-squared (R2): {r2:.2f}")

Mean Absolute Error (MAE): 8.49
Mean Squared Error (MSE): 113.68
R-squared (R2): -1.47


## Summary:

### Data Analysis Key Findings

*   The data was successfully split into training and testing sets using an 80/20 ratio.
*   A Random Forest Regressor model was trained on the training data.
*   The model's performance on the testing data was evaluated using Mean Absolute Error (MAE), Mean Squared Error (MSE), and R-squared (R2).
*   The evaluation metrics were: MAE: 8.49, MSE: 113.68, and R2: -1.47.

### Insights or Next Steps

*   The negative R-squared value suggests the current model performs worse than a simple horizontal line, indicating a need for model improvement.
*   Further steps should involve hyperparameter tuning of the Random Forest model, exploring alternative regression models, or performing feature engineering to potentially improve predictive performance.


# Task
Analyze the data from "data/content/sample_data/MSCI151 coursework 4 Data.new.xlsx" by building a predictive model using a Decision Tree Regressor. The target variable is 'Profitability', and the features are 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO Education', 'CEO Gender', 'Marketing', 'Product_sentiment', and 'PoB'. Split the data into training and testing sets using an 80/20 ratio.

## Prepare data

### Subtask:
Define the features (X) and the target variable (y) from the new DataFrame `df_new`, including the additional columns 'Marketing', 'Product_sentiment', and 'PoB'.


**Reasoning**:
Define the feature columns and the target variable from the new dataframe.



In [10]:
feature_cols_new = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing']
X_new = df_new[feature_cols_new]
y_new = df_new['Profitability']

## Split data

### Subtask:
Split the data into training and testing sets using an 80/20 rule.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [20]:
from sklearn.model_selection import train_test_split

X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X_new, y_new, test_size=0.2, random_state=42)

## Random Forest Feature Importance

### Subtask:
Display the feature importance scores for the trained Random Forest Regressor model.

**Reasoning**:
Access the feature importances from the Random Forest model and display them.

In [30]:
# Get feature importances from the Random Forest model
feature_importances_rf = model.feature_importances_

# Create a pandas Series for better visualization
feature_importance_series_rf = pd.Series(feature_importances_rf, index=X.columns)

# Sort the features by importance
sorted_feature_importances_rf = feature_importance_series_rf.sort_values(ascending=False)

# Print the feature importance scores
print("Random Forest Feature Importances:")
print(sorted_feature_importances_rf)

Random Forest Feature Importances:
debt              0.324213
Firm Size         0.164459
CEO Gender        0.156189
RnD               0.126380
Firm Age          0.070566
foreign sales     0.070101
CEO Experience    0.056322
CEO education     0.031770
dtype: float64


In [36]:
import statsmodels.api as sm
import pandas as pd

# Define the features (X) and the target variable (y)
feature_cols_ols = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product_sentiment', 'PoB']
X_ols = df_new[feature_cols_ols]
y_ols = df_new['Profitability']

# Add a constant (intercept) to the features
X_ols = sm.add_constant(X_ols)

# Create and fit the OLS model with robust standard errors
model_ols = sm.OLS(y_ols, X_ols)
results_ols = model_ols.fit(cov_type='HC1') # HC1 is a common type of robust standard error

# Print the summary of the regression results
print(results_ols.summary())

                            OLS Regression Results                            
Dep. Variable:          Profitability   R-squared:                       0.284
Model:                            OLS   Adj. R-squared:                  0.212
Method:                 Least Squares   F-statistic:                     4.859
Date:                Wed, 08 Oct 2025   Prob (F-statistic):           4.35e-06
Time:                        05:32:51   Log-Likelihood:                -465.15
No. Observations:                 122   AIC:                             954.3
Df Residuals:                     110   BIC:                             988.0
Df Model:                          11                                         
Covariance Type:                  HC1                                         
                        coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 1.6610     21.57

In [31]:
from sklearn.linear_model import Lasso

# Initialize and fit the LASSO model
# You might need to tune the alpha parameter based on your data
lasso = Lasso(alpha=0.1) # alpha is the regularization strength
lasso.fit(X_new, y_new)

# Get the coefficients
lasso_coefficients = pd.Series(lasso.coef_, index=X_new.columns)

# Print the coefficients
print("LASSO Coefficients:")
print(lasso_coefficients)

LASSO Coefficients:
debt                -0.189645
RnD                  0.024599
Firm Age            -2.482502
Firm Size            0.491453
CEO Experience       1.945461
foreign sales       -0.020908
CEO education        2.205056
CEO Gender           9.904784
Marketing           -3.889431
Product_sentiment   -0.000000
PoB                 -0.000000
dtype: float64


## Split data

### Subtask:
Split the data into training and testing sets using an 80/20 rule.

**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.

In [22]:
from sklearn.model_selection import train_test_split

X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X_new, y_new, test_size=0.2, random_state=42)

## Prepare data

### Subtask:
Define the features (X) and the target variable (y) from the new DataFrame `df_new`, including the additional columns 'Marketing', 'Product_sentiment', and 'PoB'.

**Reasoning**:
Define the feature columns and the target variable from the new dataframe.

In [21]:
feature_cols_new = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product_sentiment', 'PoB']
X_new = df_new[feature_cols_new]
y_new = df_new['Profitability']

## Evaluate model

### Subtask:
Evaluate the trained model on the testing data using appropriate metrics.


**Reasoning**:
Import necessary evaluation metrics, make predictions on the test set, calculate the metrics, and print them.



In [15]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

y_pred_new = model_dt.predict(X_test_new)

mae_new = mean_absolute_error(y_test_new, y_pred_new)
mse_new = mean_squared_error(y_test_new, y_pred_new)
r2_new = r2_score(y_test_new, y_pred_new)

print(f"Mean Absolute Error (MAE) [New Data]: {mae_new:.2f}")
print(f"Mean Squared Error (MSE) [New Data]: {mse_new:.2f}")
print(f"R-squared (R2) [New Data]: {r2_new:.2f}")

Mean Absolute Error (MAE) [New Data]: 8.26
Mean Squared Error (MSE) [New Data]: 151.61
R-squared (R2) [New Data]: -2.29


## Summary:

### Data Analysis Key Findings

*   The Decision Tree Regressor model, when evaluated on the testing data, yielded a Mean Absolute Error (MAE) of 8.26, a Mean Squared Error (MSE) of 151.61, and an R-squared (R2) score of -2.29.

### Insights or Next Steps

*   The negative R2 score suggests that the current Decision Tree Regressor model performs worse than a simple baseline model that predicts the average profitability.
*   Further investigation is needed to understand the reasons for the negative R2 score. This could involve exploring alternative models, feature engineering, or analyzing the data for potential issues or outliers.


# Task
Analyze the dataset "data/content/sample_data/MSCI151 coursework 4 Data.new.xlsx" by performing predictive modeling using Decision Tree, Random Forest, and XGBoost algorithms. For each model, use 'Profitability' as the target variable (Y) and 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO Education', 'CEO Gender', 'marketing', 'product sentiment', and 'PoB' as the predictor variables (X). Split the data into training and testing sets using an 80/20 rule.

## Prepare data

### Subtask:
Define the features (X) and the target variable (y) from the new DataFrame `df_new`, including the additional columns 'Marketing', 'Product_sentiment', and 'PoB'.


**Reasoning**:
Define the feature columns and the target variable from the new dataframe.



In [23]:
feature_cols_new = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product_sentiment', 'PoB']
X_new = df_new[feature_cols_new]
y_new = df_new['Profitability']

## Split data

### Subtask:
Split the data into training and testing sets using an 80/20 rule.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [24]:
from sklearn.model_selection import train_test_split

X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X_new, y_new, test_size=0.2, random_state=42)

# Task
Perform predictive modeling using a Random Forest model with 'Profitability' as the target variable and 'debt', 'RnD', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', and 'Marketing' as features, using an 80/20 train/test split on the data from "data/content/sample_data/MSCI151 coursework 4 Data.new.xlsx".

## Prepare data

### Subtask:
Define the features (X) and the target variable (y) from the DataFrame `df_new`, using the specified feature columns.


**Reasoning**:
Define the feature columns and the target variable from the new dataframe for the Random Forest model.



In [32]:
feature_cols_rf = ['debt', 'RnD', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing']
X_rf = df_new[feature_cols_rf]
y_rf = df_new['Profitability']

## Split data

### Subtask:
Split the data into training and testing sets using an 80/20 rule.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [33]:
from sklearn.model_selection import train_test_split

X_train_rf, X_test_rf, y_train_rf, y_test_rf = train_test_split(X_rf, y_rf, test_size=0.2, random_state=42)

## Train model

### Subtask:
Train a new Random Forest Regressor model on the training data.


**Reasoning**:
Train a Random Forest Regressor model on the training data.



In [34]:
from sklearn.ensemble import RandomForestRegressor

model_rf = RandomForestRegressor(n_estimators=100, random_state=42)
model_rf.fit(X_train_rf, y_train_rf)

RandomForestRegressor(random_state=42)

## Evaluate model

### Subtask:
Evaluate the trained model on the testing data using appropriate metrics.


**Reasoning**:
Import necessary evaluation metrics, make predictions on the test set, calculate the metrics, and print them.



In [35]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

y_pred_rf = model_rf.predict(X_test_rf)

mae_rf = mean_absolute_error(y_test_rf, y_pred_rf)
mse_rf = mean_squared_error(y_test_rf, y_pred_rf)
r2_rf = r2_score(y_test_rf, y_pred_rf)

print(f"Mean Absolute Error (MAE) [Random Forest New Data]: {mae_rf:.2f}")
print(f"Mean Squared Error (MSE) [Random Forest New Data]: {mse_rf:.2f}")
print(f"R-squared (R2) [Random Forest New Data]: {r2_rf:.2f}")

Mean Absolute Error (MAE) [Random Forest New Data]: 8.89
Mean Squared Error (MSE) [Random Forest New Data]: 132.26
R-squared (R2) [Random Forest New Data]: -1.87


## Summary:

### Data Analysis Key Findings
*   The Random Forest model trained to predict 'Profitability' using the specified features resulted in a Mean Absolute Error (MAE) of 8.89 on the test set.
*   The Mean Squared Error (MSE) for the model on the test set was 132.26.
*   The R-squared (R2) score for the model was -1.87, suggesting that the model performs significantly worse than a simple model that predicts the mean of the target variable.

### Insights or Next Steps
*   The negative R-squared score indicates that the current Random Forest model is not suitable for predicting 'Profitability' with the chosen features and data split. Further investigation is needed to understand the reasons for this poor performance.
*   Explore alternative regression models, feature engineering techniques, or hyperparameter tuning for the Random Forest model to improve predictive performance.


# Task
Perform a Random Forest regression on the dataset "data/content/sample_data/MSCI151 coursework 4 Data.new.xlsx" to predict 'Profitability' using 'debt', 'CEO Gender', and 'Marketing' as features. Split the data into 80% training and 20% testing sets and display the feature importance scores.

## Prepare data

### Subtask:
Define the features (X) using 'debt', 'CEO Gender', and 'Marketing' and the target variable (y) as 'Profitability' from the `df_new` DataFrame.


**Reasoning**:
Define the features and target variable for the Random Forest model using the specified columns.



In [37]:
selected_features = ['debt', 'CEO Gender', 'Marketing']
X_selected = df_new[selected_features]
y_selected = df_new['Profitability']

## Split data

### Subtask:
Split the data into training and testing sets using an 80/20 rule.


**Reasoning**:
Import the train_test_split function and split the data into training and testing sets.



In [38]:
from sklearn.model_selection import train_test_split

X_train_selected, X_test_selected, y_train_selected, y_test_selected = train_test_split(X_selected, y_selected, test_size=0.2, random_state=42)

## Train model

### Subtask:
Train a Random Forest Regressor model on the training data.


**Reasoning**:
Train a Random Forest Regressor model on the training data.



In [39]:
from sklearn.ensemble import RandomForestRegressor

model_selected_features = RandomForestRegressor(n_estimators=100, random_state=42)
model_selected_features.fit(X_train_selected, y_train_selected)

RandomForestRegressor(random_state=42)

## Evaluate model

### Subtask:
Evaluate the trained model on the testing data using appropriate metrics.


**Reasoning**:
Import necessary evaluation metrics, make predictions on the test set, calculate the metrics, and print them.



In [40]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

y_pred_selected = model_selected_features.predict(X_test_selected)

mae_selected = mean_absolute_error(y_test_selected, y_pred_selected)
mse_selected = mean_squared_error(y_test_selected, y_pred_selected)
r2_selected = r2_score(y_test_selected, y_pred_selected)

print(f"Mean Absolute Error (MAE) [Selected Features]: {mae_selected:.2f}")
print(f"Mean Squared Error (MSE) [Selected Features]: {mse_selected:.2f}")
print(f"R-squared (R2) [Selected Features]: {r2_selected:.2f}")

Mean Absolute Error (MAE) [Selected Features]: 9.03
Mean Squared Error (MSE) [Selected Features]: 177.69
R-squared (R2) [Selected Features]: -2.86


## Display feature importance

### Subtask:
Show the feature importance scores for this specific Random Forest model.


**Reasoning**:
Access the feature importances from the Random Forest model and display them.



In [41]:
# Get feature importances from the Random Forest model
feature_importances_selected = model_selected_features.feature_importances_

# Create a pandas Series for better visualization
feature_importance_series_selected = pd.Series(feature_importances_selected, index=selected_features)

# Sort the features by importance
sorted_feature_importances_selected = feature_importance_series_selected.sort_values(ascending=False)

# Print the feature importance scores
print("Random Forest Feature Importances (Selected Features):")
print(sorted_feature_importances_selected)

Random Forest Feature Importances (Selected Features):
debt          0.566176
Marketing     0.255524
CEO Gender    0.178300
dtype: float64


## Principal Component Analysis

### Subtask:
Perform PCA on the specified features and display the loading factors.

**Reasoning**:
Import necessary libraries, standardize the features, perform PCA, and display the loading factors.

In [42]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import pandas as pd

# Define the features for PCA
feature_cols_pca = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'Marketing', 'Product_sentiment', 'PoB']
X_pca = df_new[feature_cols_pca]

# Standardize the features
scaler = StandardScaler()
X_pca_scaled = scaler.fit_transform(X_pca)

# Perform PCA
pca = PCA()
X_pca_transformed = pca.fit_transform(X_pca_scaled)

# Get the loading factors
loading_factors = pd.DataFrame(pca.components_.T, columns=[f'PC{i+1}' for i in range(pca.n_components_)], index=feature_cols_pca)

# Display the loading factors
print("PCA Loading Factors:")
print(loading_factors)

PCA Loading Factors:
                        PC1       PC2       PC3       PC4       PC5       PC6  \
debt              -0.137278 -0.264457  0.022901  0.199147  0.722276  0.296221   
RnD                0.248985  0.035718 -0.381083 -0.424983  0.306896  0.225746   
Firm Age           0.404094  0.187400 -0.206086  0.427116  0.057615 -0.193697   
Firm Size          0.296997  0.307233  0.013418  0.239870 -0.011418  0.430006   
CEO Experience     0.142157  0.662708 -0.199195  0.106621  0.154872 -0.057999   
foreign sales     -0.204697  0.382921  0.396491 -0.280492  0.082339 -0.163574   
CEO education      0.516262 -0.285947  0.300485 -0.088410  0.049326  0.262396   
CEO Gender         0.487960  0.037835  0.408698 -0.311446 -0.142420 -0.031440   
Marketing          0.010582 -0.078730  0.393992  0.584490 -0.075232 -0.015253   
Product_sentiment -0.189455  0.320349  0.446839 -0.056394  0.362265  0.140669   
PoB               -0.254785  0.146446 -0.049956  0.009843 -0.436239  0.720339   

      

## Summary:

### Data Analysis Key Findings

*   The Random Forest model trained with 'debt', 'CEO Gender', and 'Marketing' as features resulted in a Mean Absolute Error (MAE) of 9.03, a Mean Squared Error (MSE) of 177.69, and an R-squared (R2) score of -2.86 on the test set.
*   The feature importance scores indicate that 'debt' is the most important feature for predicting 'Profitability' (approximately 56.6% importance), followed by 'Marketing' (approximately 25.6%), and 'CEO Gender' (approximately 17.8%).

### Insights or Next Steps

*   The negative R2 score suggests that the selected features and the current model configuration do not perform well in predicting 'Profitability'. Future steps should involve exploring additional relevant features or alternative regression models.
*   Investigating the distribution and potential interactions between 'debt', 'Marketing', and 'Profitability' could provide further insights into their relationships and potentially improve model performance.
